pkg/cmd/server/defaults.go

```diff
@@ -139,109 +139,171 @@ const (
 // DefaultUnaryMiddleware generates the default middleware chain used for the public SpiceDB Unary gRPC methods
 func DefaultUnaryMiddleware(logger zerolog.Logger, authFunc grpcauth.AuthFunc, enableVersionResponse bool, dispatcher dispatch.Dispatcher, ds datastore.Datastore) (*MiddlewareChain[grpc.UnaryServerInterceptor], error) {
 	chain, err := NewMiddlewareChain([]ReferenceableMiddleware[grpc.UnaryServerInterceptor]{
-		{
-			Name:       DefaultMiddlewareRequestID,
-			Middleware: requestid.UnaryServerInterceptor(requestid.GenerateIfMissing(true)),
-		},
-		{
-			Name:       DefaultMiddlewareLog,
-			Middleware: logmw.UnaryServerInterceptor(logmw.ExtractMetadataField("x-request-id", "requestID")),
-		},
-		{
-			Name:       DefaultMiddlewareGRPCLog,
-			Middleware: grpclog.UnaryServerInterceptor(grpczerolog.InterceptorLogger(logger), defaultGRPCLogOptions...),
-		},
-		{
-			Name:       DefaultMiddlewareOTelGRPC,
-			Middleware: otelgrpc.UnaryServerInterceptor(),
-		},
-		{
-			Name:       DefaultMiddlewareGRPCProm,
-			Middleware: grpcprom.UnaryServerInterceptor,
-		},
-		{
-			Name:       DefaultMiddlewareGRPCAuth,
-			Middleware: grpcauth.UnaryServerInterceptor(authFunc),
-		},
-		{
-			Name:       DefaultMiddlewareServerVersion,
-			Middleware: serverversion.UnaryServerInterceptor(enableVersionResponse),
-		},
-		{
-			Name:       DefaultInternalMiddlewareDispatch,
-			Internal:   true,
-			Middleware: dispatchmw.UnaryServerInterceptor(dispatcher),
-		},
-		{
-			Name:       DefaultInternalMiddlewareDatastore,
-			Internal:   true,
-			Middleware: datastoremw.UnaryServerInterceptor(ds),
-		},
-		{
-			Name:       DefaultInternalMiddlewareConsistency,
-			Internal:   true,
-			Middleware: consistencymw.UnaryServerInterceptor(),
-		},
-		{
-			Name:       DefaultInternalMiddlewareServerSpecific,
-			Internal:   true,
-			Middleware: servicespecific.UnaryServerInterceptor,
-		},
+		NewUnaryMiddleware().
+			WithName(DefaultMiddlewareRequestID).
+			WithInterceptor(requestid.UnaryServerInterceptor(requestid.GenerateIfMissing(true))).
+			EnsureAlreadyExecuted(none).
+			EnsureNotExecuted(DefaultMiddlewareLog).
+			Done(),
+
+		NewUnaryMiddleware().
+			WithName(DefaultMiddlewareLog).
+			WithInterceptor(logmw.UnaryServerInterceptor(logmw.ExtractMetadataField("x-request-id", "requestID"))).
+			EnsureAlreadyExecuted(DefaultMiddlewareRequestID).
```

Comment:
I think we should have the Ensure calls when its necessary; if a middleware doesn't care about ordering relative to other middlewares, no need for the check

Reply:
Do you know if any of these middlewares care about ordering?